### PR TITLE
exit instead of traceback when directory is no borg repository

### PR DESCRIPTION
**Pull request by tehtbl**
Hi!

here is a fix for #4764 :)

Please don't hesitate to contact me in case of open questions or advise for improvements.

Thanks,
Thomas

### CODE REVIEW

**Comment by witten** [comment]
Hooray!

File: src/borg/repository.py [hunk 1]
```diff
@@ -396,7 +396,8 @@ def open(self, path, exclusive, lock_wait=None, lock=True):
         try:
             st = os.stat(path)
         except FileNotFoundError:
-            raise self.DoesNotExist(path)
+            logger.warning("%s is not a valid borgbackup directory." % (path))
+            os.exit(2)
```
Comment (ThomasWaldmann) [inline src/borg/repository.py]: this does not seem to fix the issue you referred to, but it changes the behaviour if there is no repository directory at all.

not sure the behaviour change is the right way either, see the other places below where it handles misc. error conditions also by raising a specific exception, so why should it be different here?

in general, you need to reproduce the issue without any change, then fix it and try to reproduce it again (or even better: add a test, so we can automatically test that).

for repository code changes, you also need to test with a local repo and a ssh: repo (borg in client/server mode) and make sure both scenarious work ok.

**Comment by tehtbl** [comment]
Hi @ThomasWaldmann 

thank you for your suggestions! I've updated the PR accordingly. 

I have reproduced the issue as reported in the issue itself via `mkdir empty-dir ; borg info empty-dir` while being in the Vagrant environment. There, I've also did my tests.

I am not fluent in writing tests yet. But if you want any test case for this PR please let me know and I will create one.

thanks,
Thomas

File: src/borg/repository.py [hunk 2]
```diff
@@ -396,8 +396,8 @@ def open(self, path, exclusive, lock_wait=None, lock=True):
         try:
             st = os.stat(path)
         except FileNotFoundError:
-            logger.warning("%s is not a valid borgbackup directory." % (path))
-            os.exit(2)
+            self.close()
```
Comment (ThomasWaldmann) [inline src/borg/repository.py]: did you look at what `close()` is doing and whether it is needed here?

File: src/borg/repository.py [hunk 3]
```diff
@@ -396,7 +396,8 @@ def open(self, path, exclusive, lock_wait=None, lock=True):
         try:
             st = os.stat(path)
         except FileNotFoundError:
-            raise self.DoesNotExist(path)
+            self.close()
+            raise self.InvalidRepository(self.path)
```
Comment (ThomasWaldmann) [inline src/borg/repository.py]: overall, you changed the "does not exist" (which is true) into a "invalid" (which is questionable) exception.

what's the point?

Comment (ThomasWaldmann) [reply src/borg/repository.py]: also, please do not merge master into your branch if there is no reason to do so.

**Comment by tehtbl** [comment]
I re-checked all about the issue again. Long story short, the PR #4461 already fixed the issue. The Bug is still valid in version 1.1.10 but no longer in the current master branch (1.2.0a8.dev9+g73ab1f1d.d20191007).

Sorry for taking your time :/

**Comment by ThomasWaldmann** [comment]
@tehtbl oh, interesting find - thanks for noting this.

i'll have a look whether we want to backport #4461.

**Pull request closed without merging**